File: crates/livekit-webrtc/libwebrtc-sys/src/data_channel.cpp
//
// Created by Théo Monnom on 01/09/2022.
//

#include <utility>

#include "livekit/data_channel.h"
#include "libwebrtc-sys/src/data_channel.rs.h"

namespace livekit {

    DataChannel::DataChannel(rtc::scoped_refptr<webrtc::DataChannelInterface> data_channel) : data_channel_(std::move(data_channel)) {

    }

    void DataChannel::register_observer(std::unique_ptr<NativeDataChannelObserver> observer) {
        data_channel_->RegisterObserver(observer.get());
    }

    void DataChannel::unregister_observer() {
        data_channel_->UnregisterObserver();
    }

    void DataChannel::close() {
        return data_channel_->Close();
    }

    std::unique_ptr<NativeDataChannelInit> create_data_channel_init(DataChannelInit init) {
        auto rtc_init = std::make_unique<webrtc::DataChannelInit>();
        rtc_init->id = init.id;
        rtc_init->negotiated = init.negotiated;
        rtc_init->ordered = init.ordered;
        rtc_init->protocol = init.protocol.c_str();
        rtc_init->reliable = init.reliable;

        if(init.has_max_retransmit_time)
            rtc_init->maxRetransmitTime = init.max_retransmit_time;

        if(init.has_max_retransmits)
            rtc_init->maxRetransmits = init.max_retransmits;

        if(init.has_priority)
            rtc_init->priority = static_cast<webrtc::Priority>(init.priority);

        return rtc_init;
    }

    NativeDataChannelObserver::NativeDataChannelObserver(rust::Box<DataChannelObserverWrapper> observer) : observer_(std::move(observer)){

    }

    void NativeDataChannelObserver::OnStateChange() {
        observer_->on_state_change();
    }

    void NativeDataChannelObserver::OnMessage(const webrtc::DataBuffer &buffer) {
        DataBuffer data{};
        data.binary = buffer.data.data();
        data.len = buffer.data.size();
        data.binary = buffer.binary;
        observer_->on_message(data);
    }

    void NativeDataChannelObserver::OnBufferedAmountChange(uint64_t sent_data_size) {
        observer_->on_buffered_amount_change(sent_data_size);
    }

    std::unique_ptr<NativeDataChannelObserver> create_native_peer_connection_observer(rust::Box<DataChannelObserverWrapper> observer){
        return std::make_unique<NativeDataChannelObserver>(std::move(observer));
    }
} // livekit

File: crates/livekit-webrtc/libwebrtc-sys/src/data_channel.rs
use cxx::UniquePtr;
use std::slice;

#[cxx::bridge(namespace = "livekit")]
pub mod ffi {

    #[derive(Debug)]
    #[repr(u32)]
    pub enum Priority {
        VeryLow,
        Low,
        Medium,
        High,
    }

    #[derive(Debug)]
    pub struct DataChannelInit {
        #[allow(deprecated)]
        #[deprecated]
        pub reliable: bool,
        pub ordered: bool,
        pub has_max_retransmit_time: bool,
        pub max_retransmit_time: i32,
        pub has_max_retransmits: bool,
        pub max_retransmits: i32,
        pub protocol: String,
        pub negotiated: bool,
        pub id: i32,
        pub has_priority: bool,
        pub priority: Priority,
    }

    #[derive(Debug)]
    pub struct DataBuffer {
        pub ptr: *const u8,
        pub len: usize,
        pub binary: bool,
    }

    #[derive(Debug)]
    pub enum DataState {
        Connecting,
        Open,
        Closing,
        Closed,
    }

    extern "Rust" {
        type DataChannelObserverWrapper;

        fn on_state_change(self: &DataChannelObserverWrapper);
        fn on_message(self: &DataChannelObserverWrapper, buffer: DataBuffer);
        fn on_buffered_amount_change(self: &DataChannelObserverWrapper, sent_data_size: u64);
    }

    unsafe extern "C++" {
        include!("livekit/data_channel.h");

        type DataChannel;
        type NativeDataChannelInit;
        type NativeDataChannelObserver;

        /// SAFETY
        /// The observer must live as the datachannel uses it
        unsafe fn register_observer(
            self: Pin<&mut DataChannel>,
            observer: Pin<&mut NativeDataChannelObserver>,
        );

        fn unregister_observer(self: Pin<&mut DataChannel>);
        fn close(self: Pin<&mut DataChannel>);

        fn create_data_channel_init(init: DataChannelInit) -> UniquePtr<NativeDataChannelInit>;
        fn create_native_data_channel_observer(
            observer: Box<DataChannelObserverWrapper>,
        ) -> UniquePtr<NativeDataChannelObserver>;

        fn _unique_data_channel() -> UniquePtr<DataChannel>; // Ignore
    }
}

// DataChannelObserver

pub trait DataChannelObserver: Send {
    fn on_state_change(&self);
    fn on_message(&self, data: &[u8], is_binary: bool);
    fn on_buffered_amount_change(&self, sent_data_size: u64);
}

pub struct DataChannelObserverWrapper {
    observer: Box<dyn DataChannelObserver>,
}

impl DataChannelObserverWrapper {
    pub fn new(observer: Box<dyn DataChannelObserver>) -> Self {
        Self { observer }
    }

    fn on_state_change(&self) {
        self.observer.on_state_change();
    }

    fn on_message(&self, buffer: ffi::DataBuffer) {
        let data = unsafe { slice::from_raw_parts(buffer.ptr, buffer.len) };
        self.observer.on_message(data, buffer.binary);
    }

    fn on_buffered_amount_change(&self, sent_data_size: u64) {
        self.observer.on_buffered_amount_change(sent_data_size);
    }
}
